File: python/euler/bin/p0037.py
# project euler: problem 37

# candidate numbers: [2357][1379]*[37] (n >= 10)
#                          ----------- lst

from euler.lib.prime import is_prime
from euler.lib.util import num_of_digits


def add_prefix_num(pre_lst: list[int], lst: list[int]) -> list[int]:
    return [p * (10 ** num_of_digits(n)) + n for p in pre_lst for n in lst]


def make_next_lists(lst: list[int]) -> list[int]:
    return list(filter(is_prime, add_prefix_num([1, 3, 7, 9], lst)))


def pickup_primes(lst: list[int]) -> list[int]:
    def is_truncable_prime(n: int) -> bool:
        if n == 0:
            return False
        while n != 0:
            if not is_prime(n):
                return False
            n //= 10
        return True

    return list(filter(is_truncable_prime, add_prefix_num([2, 3, 5, 7], lst)))


def compute() -> str:
    result: list[int] = []
    lst = [3, 7]
    while len(result) < 11:
        result += pickup_primes(lst)
        lst = make_next_lists(lst)
    assert len(result) == 11, 'many candidates'

    return str(sum(result))


def solve() -> str:
    return compute()

File: typescript/euler/bin/p0037.ts
// project euler: problem 37

/*
  candidate numbers: [2357][1379]*[37]
                           ----------- lst
*/

import { sum } from "../lib/math.ts";
import { isPrime } from "../lib/prime.ts";
import { numOfDigits } from "../lib/util.ts";

const addPrefix = (prefixes: number[], lst: number[]): number[] => {
  const acc: number[] = [];
  for (const n of lst) {
    for (const p of prefixes) {
      acc.push(p * (10 ** numOfDigits(n)) + n);
    }
  }

  return acc;
};

const makeNextLst = (lst: number[]): number[] =>
  addPrefix([1, 3, 7, 9], lst).filter((x) => isPrime(x));

const pickupPrimes = (lst: number[]): number[] => {
  const isTruncablePrime = (n: number): boolean => {
    if (n === 0) {
      return false;
    }
    while (n !== 0) {
      if (!isPrime(n)) {
        return false;
      }
      n = Math.trunc(n / 10);
    }

    return true;
  };

  return addPrefix([2, 3, 5, 7], lst).filter((x) => isTruncablePrime(x));
};

export const compute = (): string => {
  let result: number[] = [];
  let lst = [3, 7];
  while (result.length < 11) {
    result = result.concat(pickupPrimes(lst));
    lst = makeNextLst(lst);
  }

  return String(sum(result));
};

export const solve = (): string => compute();
